### PR TITLE
Avoid __CUDA_ARCH__ in source code

### PR DESCRIPTION
**Pull request by masterleinad**
I really don't like also the duplication for `Kokkos` versions earlier than 3.6.00.

### CODE REVIEW

**Review by drwells — APPROVED** [review]
Nice!

File: include/deal.II/base/exceptions.h [hunk 1]
```diff
@@ -28,6 +28,9 @@
 #  include <cusparse.h>
 #endif
 
+DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
+#include <Kokkos_Core.hpp>
+DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
```
Comment (drwells) [inline include/deal.II/base/exceptions.h:33]: This file, when preprocessed, is 100k lines - that's a lot to add to every single file which includes `exceptions.h`. Can we get away with doing something more minimal?

Comment (drwells) [reply include/deal.II/base/exceptions.h:33]: I didn't try it but maybe `Kokkos_Macros.hpp` is enough.

Comment (masterleinad) [reply include/deal.II/base/exceptions.h:33]: We need `Kokkos_Core.hpp` for `Kokkos::abort`. We could define another (device-callable) function taking a `const char *` to wrap `Kokkos::abort` that we use and put its definition in a source file. We certainly don't care about it being inlined anyway.

File: include/deal.II/base/exceptions.h [hunk 2]
```diff
@@ -1505,39 +1506,91 @@ namespace deal_II_exceptions
...
+                __LINE__,                                                    \
+                __PRETTY_FUNCTION__,                                         \
+                #cond,                                                       \
+                #exc,                                                        \
+                exc);                                                        \
+          }))                                                                \
+          KOKKOS_IF_ON_DEVICE(({                                             \
+            if (!(cond))                                                     \
+              dealii::internal::kokkos_abort(#cond);                         \
+          }))                                                                \
```
Comment (drwells) [inline include/deal.II/base/exceptions.h:1528]: Another question - can we merge `internal::kokkos_abort()` and `issue_error_noreturn()`? What is preventing us from moving `KOKKOS_IF_ON_HOST()` into that function?

Comment (masterleinad) [reply include/deal.II/base/exceptions.h:1528]: The exception variables can't be built on GPUs since they are using `std::string` and virtual functions.

**Review by tamiko — APPROVED** [review]

**Pull request merged**